### ISSUE TITLE
URLSearchParams breaks ISO8601 date strings 

### ISSUE THREAD

**Issue by marvinscharle** [angular/angular#13077]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report 
[ ] feature request
[ ] support request
```

**Current behavior**
When encoding `URLSearchParams`, plus signs are not correctly encoded and remain `+` instead of `&2B`. This leads to plus signs being transmitted as space chars. 

Example: 

```typescript
let params = new URLSearchParams();
params.set('date', '2016-11-24 12:11:00+01:00');

params.toString(); // => 'date=2016-11-24%2012:11:00+01:00'
```

In the example, the date param would now contain an invalid ISO8601 date string (`2016-11-24 12:11:00 01:00` - plus was interpreted as space).

**Expected behavior**
Plus signs should be replaced with `&2B` so that they remain plus signs.

**Minimal reproduction of the problem with instructions**
http://plnkr.co/edit/FjxyBJlcaoTs8k98X6sN

**What is the motivation / use case for changing the behavior?**
Minimal use case is to transfer ISO8601 encoded Date objects.

**Please tell us about your environment:**
macOS 10.12, Google Chrome 54

* **Angular version:** 
Latest, 2.1.X, 2.2.X
 
* **Language:** 
Typescript

**Comment by skreborn** [comment]
I suppose the culprit is [this](https://github.com/angular/angular/blob/8f5dd1f11e6ca1888fdbd3231c06d6df00aba5cc/modules/%40angular/http/src/url_search_params.ts#L40) line in `http/src/url_search_params`.

The `standardEncoding` function returns the desired string (`2016-11-24%2012:11:00%2B01:00`) when the line is omitted.
```
function standardEncoding(v) {
  return encodeURIComponent(v)
      .replace(/%40/gi, '@')
      .replace(/%3A/gi, ':')
      .replace(/%24/gi, '$')
      .replace(/%2C/gi, ',')
      .replace(/%3B/gi, ';')
//    .replace(/%2B/gi, '+')
      .replace(/%3D/gi, '=')
      .replace(/%3F/gi, '?')
      .replace(/%2F/gi, '/');
}

standardEncoding('2016-11-24 12:11:00+01:00');
```

**Comment by DzmitryShylovich** [comment]
> If the set of allowed query characters is not acceptable for a particular backend, `QueryEncoder` can be subclassed and provided as the 2nd argument to URLSearchParams.

from the docs. They intentionally un encode `+`.

**Comment by ericmartinezr** [comment]
https://github.com/angular/angular/issues/11058

**Comment by alxhub** [comment]
I will look into why `+` is not encoded by default.

**Comment by tiangolo** [comment]
I created a PR to solve this here: https://github.com/angular/angular/pull/19711

And to solve the same problem when using `HttpClientModule` and `HttpClient`: https://github.com/angular/angular/pull/19710 .

But they have been there waiting to be reviewed since October.

---

So, I created a package to solve it (at least while it is reviewed / merged) for the case of `HttpClient`. It's a simple drop-in replacement that modifies the minimum to fix the problem. It works with just changing one import.

Check it here: https://github.com/tiangolo/ngx-http-client

**Comment by alxhub** [comment]
Closing as this is obsolete (it concerns `@angular/http` which has been removed in favor of `@angular/common/http`).

For `HttpClient`, I believe the bug was fixed but #11058 also tracks work we need to do with encoding of parameters in general.
